### PR TITLE
[SPARK-45713][PYTHON] Support registering Python data sources

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

This PR adds support for registering Python data sources.
Users can register a Python data source using the class:

```python
class MyDataSource(DataSource):
    ...

spark.dataSource.register(MyDataSource)
```

This will allow users to use the data source using its name (to be supported in SPARK-45639)
```python
spark.read.format("MyDataSource").load()
```

The data sources registered are stored in `sharedState` and can be accessed by all spark sessions.

### Why are the changes needed?

To support Python data source.

### Does this PR introduce _any_ user-facing change?

Yes. This PR adds a new API in spark session to support registering data sources.

### How was this patch tested?

Unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by allisonwang-db** [comment]
cc @HyukjinKwon

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceRegistry.scala [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.errors.QueryCompilationErrors
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+
+trait DataSourceRegistry {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceRegistry.scala:33]: I think we should somehow reuse `DataSourceRegister`.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceRegistry.scala:33]: cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/DataSourceRegistration.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import org.apache.spark.annotation.Evolving
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.execution.datasources.DataSourceRegistry
+import org.apache.spark.sql.execution.python.UserDefinedPythonDataSource
+
+/**
+ * Functions for registering user-defined data sources.
+ * Use `SparkSession.dataSource` to access this.
+ *
+ * @since 4.0.0
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataSourceRegistration.scala:29]: do we need to expose it now? It's only called by PySpark internally. We can mark it as `private[sql]`

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 3]
```diff
@@ -228,6 +228,11 @@ class SparkSession private(
 
   def udtf: UDTFRegistration = sessionState.udtfRegistration
 
+  /**
+   * A collection of methods for registering user-defined data sources.
+   */
+  def dataSource: DataSourceRegistration = sharedState.dataSourceRegistration
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:234]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceRegistry.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceRegistry.scala:33]: These two are completely different. We should probably rename this one here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceRegistry.scala [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+  def registerDataSource(name: String, builder: DataSourceBuilder): Unit
+
+  def dataSourceExists(name: String): Boolean
+
+  /** Create a copy of this registry with identical data sources as this registry. */
+  override def clone(): DataSourceRegistry = throw new CloneNotSupportedException()
+}
+
+class SimpleDataSourceRegistry extends DataSourceRegistry with Logging {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceRegistry.scala:51]: Let's not add unnecessary abstraction. Why would we need different implementations of `DataSourceRegistry`? Isn't it just a mapping between name and builder?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceRegistry.scala (hunk 1 above)
Comment (allisonwang-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceRegistry.scala:33]: Yea we should pick a different name. The name "registry" is taken from "FunctionRegistry" and "TableFunctionRegistry". How about "DataSourceStore" or "DataSourceManager"

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceRegistry.scala (hunk 4 above)
Comment (allisonwang-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceRegistry.scala]: Sounds good. Changed this to a simple class with a concurrent hasmap.

**Comment by allisonwang-db** [comment]
@HyukjinKwon @cloud-fan Thanks for the review! I've addressed the comments. Note this PR does not include the data source lookup logic. I will work on it in the next PR ([SPARK-45639](https://issues.apache.org/jira/browse/SPARK-45639)) (and also address any additional comments)

File: sql/core/src/main/scala/org/apache/spark/sql/DataSourceRegistration.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import org.apache.spark.annotation.Evolving
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.execution.datasources.DataSourceManager
+import org.apache.spark.sql.execution.python.UserDefinedPythonDataSource
+
+/**
+ * Functions for registering user-defined data sources.
+ * Use `SparkSession.dataSource` to access this.
+ */
+@Evolving
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataSourceRegistration.scala:29]: nit: we don't need it for private APIs.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/DataSourceRegistration.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/DataSourceRegistration.scala:29]: We can fix it in the followup

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by HyukjinKwon** [comment]
Hey, I don't think we should go ahead. It should have a lookup logic first, and then we should think about something extra runtime registration

**Comment by panbingkun** [comment]
After this pr, failed to run 'SparkThrowableSuite' in GA, eg:
https://github.com/panbingkun/spark/actions/runs/6702704679/job/18212102847
image
I have submitted a PR to fix this issue: https://github.com/apache/spark/pull/43598

**Pull request closed without merging**